test/hotspot/jtreg/compiler/lib/compile_framework/CompileFramework.java

```diff
@@ -0,0 +1,303 @@
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.concurrent.TimeUnit;
+import java.util.ArrayList;
+import java.util.List;
+
+public class CompileFramework {
+    private static final int COMPILE_TIMEOUT = 60;
+    private static final boolean VERBOSE = Boolean.getBoolean("CompileFrameworkVerbose");
+
+    private List<SourceCode> sourceCodes = new ArrayList<SourceCode>();
+    private final Path sourceDir = getTempDir("compile-framework-sources-");
+    private final Path classesDir = getTempDir("compile-framework-classes-");
+    private URLClassLoader classLoader;
+
+    public String getClassPathOfCompiledClasses() {
+        String cp = System.getProperty("java.class.path") +
+                    File.pathSeparator +
+                    classesDir.toAbsolutePath().toString();
+        return cp.replace("\\", "\\\\"); // For windows paths
+    }
+
+    public void add(SourceCode sourceCode) {
+        sourceCodes.add(sourceCode);
+    }
+
+    public String sourceCodesAsString() {
+        StringBuilder builder = new StringBuilder();
+        for (SourceCode sourceCode : sourceCodes) {
+            builder.append("SourceCode: ").append(sourceCode.filePathName()).append(System.lineSeparator());
+            builder.append(sourceCode.code).append(System.lineSeparator());
+        }
+        return builder.toString();
+    }
+
+    private static void println(String s) {
+        if (VERBOSE) {
+            System.out.println(s);
+        }
+    }
+
+    public void compile() {
+        if (classLoader != null) {
+            throw new CompileFrameworkException("Cannot compile twice!");
+        }
+
+        println("------------------ CompileFramework: SourceCode -------------------");
+        println(sourceCodesAsString());
+
+        List<SourceCode> javaSources = new ArrayList<SourceCode>();
+        List<SourceCode> jasmSources = new ArrayList<SourceCode>();
+        for (SourceCode sourceCode : sourceCodes) {
+            switch (sourceCode.kind) {
+                case SourceCode.Kind.JASM -> { jasmSources.add(sourceCode);  }
+                case SourceCode.Kind.JAVA -> { javaSources.add(sourceCode);  }
+            }
+        }
+
+        System.out.println("------------------ CompileFramework: Compilation ------------------");
+        System.out.println("Source directory: " + sourceDir.toString());
+        System.out.println("Classes directory: " + classesDir.toString());
+
+        compileJasmSources(jasmSources);
+        compileJavaSources(javaSources);
+        setUpClassLoader();
+    }
+
+    private static Path getTempDir(String prefix) {
+        try {
+            return Files.createTempDirectory(Paths.get("."), prefix);
+        } catch (Exception e) {
+            throw new InternalCompileFrameworkException("Could not set up temporary directory", e);
+        }
+    }
+
+    private void compileJasmSources(List<SourceCode> jasmSources) {
```

Comment:
The `compileJasmSources()/compileJasmFiles()` and `compileJavaSources()/compileJavaFiles()` method seem to be very specific to whether it's a Java or Jasm file. I'm wondering if we could do the following:

- Make class `SourceCode` an interface and create two implementing classes `JavaSourceCode` and `JasmSourceCode`.
- Move the `compileJasm/JavaSources()/compileJasm/JavaFiles()` to the corresponding classes (the interface method could be `compile()`).
- This allows us to get ride of the `enum` as well in `SourceCode`. The interface could look like this:
```
interface SourceCode {
    void compile();
    String code();
    String fileExtension();
    String className();
    String filePathName(); // Could probably just provide the current implementation as default
}
```
I have not thought it through completely and I might be missing some things - but might be worth a shot :-)

Reply:
The offline discussion was quite fruitful, I'll do it another way now.